lib/floe/workflow/intrinsic_function/transformer.rb

```diff
@@ -3,41 +3,189 @@
+          Transformer.process_args(args, "States.Array")
+        end
+
+        rule(:states_array_partition => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.ArrayPartition", [Array, Integer])
+          array, chunk = *args
+          raise ArgumentError, "invalid value for argument 2 to States.ArrayPartition (given #{chunk}, expected a positive Integer)" unless chunk.positive?
+
+          array.each_slice(chunk).to_a
+        end
+
+        rule(:states_array_contains => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.ArrayContains", [Array, Object])
+          array, target = *args
+
+          array.include?(target)
+        end
+
+        rule(:states_array_range => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.ArrayRange", [Integer, Integer, Integer])
+          range_begin, range_end, increment = *args
+          raise ArgumentError, "invalid value for argument 3 to States.ArrayRange (given #{increment}, expected a non-zero Integer)" if increment.zero?
+
+          (range_begin..range_end).step(increment).to_a
+        end
+
+        rule(:states_array_get_item => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.ArrayGetItem", [Array, Integer])
+          array, index = *args
+          raise ArgumentError, "invalid value for argument 2 to States.ArrayGetItem (given #{index}, expected 0 or a positive Integer)" unless index >= 0
+
+          array[index]
+        end
+
+        rule(:states_array_length => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.ArrayLength", [Array])
+          array = args.first
+
+          array.size
+        end
+
+        rule(:states_array_unique => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.ArrayUnique", [Array])
+          array = args.first
+
+          array.uniq
+        end
+
+        rule(:states_base64_encode => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.Base64Encode", [String])
+          str = args.first
+
+          require "base64"
+          Base64.strict_encode64(str).force_encoding("UTF-8")
+        end
+
+        rule(:states_base64_decode => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.Base64Decode", [String])
+          str = args.first
+
+          require "base64"
+          begin
+            Base64.strict_decode64(str)
+          rescue ArgumentError => err
+            raise ArgumentError, "invalid value for argument 1 to States.Base64Decode (#{err})"
+          end
+        end
+
+        rule(:states_hash => {:args => subtree(:args)}) do
+          args = Transformer.process_args(args(), "States.Hash", [Object, String])
+          data, algorithm = *args
+          raise NotImplementedError if data.kind_of?(Hash)
```

Comment:
Just calling this out - I can't figure out how the stepfunctions simulator converts this to get a Hash but it does, so I just marked as Not implemented for now.

Reply:
I wonder if they .to_json the hash then checksum that string?

Reply:
Oh hadn't though of checksum - I tried json, json without spaces, pretty json

Reply:
Hm yeah no variation of JSON of the hash that I tried seems to match what you have in the spec (I assume that value came from the step-functions simulator)